### PR TITLE
A return to foam (foam rework)

### LINKED ISSUE #20732: Fix/Rework Foam Event & Foam Reactions (Do not remove event)
## Description

The Foam Event/reactions should be fixed or reworked so that the event is properly working.

The solution is to get it into a state where it's working. 

Do not remove the event from code because "it does nothing"

#20716 and #19872 are examples of what not to do.

### LINKED ISSUE #20830: Antag status doesn't clear from deconverted revs
## Description

I think there's a similar issue with zombies, it's not a major issue but it does clog the admin panel when I simply want to see who the actual main antags of the round are.

### PR DESCRIPTION
**Pull request by EmoGarbage404**
## About the PR

Adds back in the ability for foam/smoke to transfer reagents, but with a few changes:
- Smoke/foam no longer transfers the entire volume of the solution in a second, instead transferring a percentage based on the duration of the foam. In layman's terms: standing in foam/smoke should never give an individual more u of a solution than was originally present.
- Modified the spreading to be more even and less horrifically jank.
- Fixed the foam dissolving animation. Also fixed when it would mispredict at the end and play multiple times.
- Adjusted the vent foam event to be less completely insane. 200u per foam??? 

Fixes #20732

## Why / Balance

chem smoke/foam is fun and it should be back

## Media

https://github.com/space-wizards/space-station-14/assets/98561806/141a220e-cae4-4e10-a721-fc583c9b9266

video of updated spreading code + animation fix

https://github.com/space-wizards/space-station-14/assets/98561806/18e6f9b2-d6ad-48c6-8614-0242d9f41936

video of the solution transfer changes.

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes

renamed SmokeSystem.Start to SmokeSystem.StartSmoke + added new parameters. These were required anyway so this is just streamlining the function.

SmokeComponent is now in shared.

**Changelog**

:cl:
- add: Foam and smoke can now transfer chemicals when inhaled. Wear internals to avoid this.
- fix: Fixed foam dissolution animation.

### CODE REVIEW

**Comment by deltanedas** [comment]
ipecac pax foam my beloved

**Comment by brainfood1183** [comment]
this include foam artifact effects?

**Comment by Myakot** [comment]
Dissolution animation looks much better, thanks!

File: Content.Server/Fluids/EntitySystems/SmokeSystem.cs [hunk 1]
```diff
@@ -22,70 +33,130 @@ namespace Content.Server.Fluids.EntitySystems;
...
+            SmokeReact(uid, smoke.SmokeEntity);
+        }
+    }
+
+    private void OnStartCollide(EntityUid uid, SmokeComponent component, ref StartCollideEvent args)
+    {
+        if (HasComp<SmokeAffectedComponent>(args.OtherEntity))
+            return;
+
+        var smokeAffected = EnsureComp<SmokeAffectedComponent>(args.OtherEntity);
```
Comment (metalgearsloth) [inline Content.Server/Fluids/EntitySystems/SmokeSystem.cs:88]: AddComp as you already checked it.

File: Content.Server/Fluids/EntitySystems/SmokeSystem.cs [hunk 2]
```diff
@@ -22,70 +33,130 @@ namespace Content.Server.Fluids.EntitySystems;
...
+
+            smoke.NextSecond += TimeSpan.FromSeconds(1);
+            SmokeReact(uid, smoke.SmokeEntity);
+        }
+    }
+
+    private void OnStartCollide(EntityUid uid, SmokeComponent component, ref StartCollideEvent args)
+    {
+        if (HasComp<SmokeAffectedComponent>(args.OtherEntity))
+            return;
```
Comment (metalgearsloth) [inline Content.Server/Fluids/EntitySystems/SmokeSystem.cs:86]: Use EntityQuery<T> for this + onendcollide

File: Content.Shared/Chemistry/Components/SmokeAffectedComponent.cs [hunk 3]
```diff
@@ -0,0 +1,23 @@
+using Robust.Shared.Serialization.TypeSerializers.Implementations.Custom;
+
+namespace Content.Shared.Chemistry.Components;
+
+/// <summary>
+/// This is used for entities which are currently being affected by smoke.
+/// Manages the gradual metabolism every second.
+/// </summary>
+[RegisterComponent]
+public sealed partial class SmokeAffectedComponent : Component
```
Comment (metalgearsloth) [inline Content.Shared/Chemistry/Components/SmokeAffectedComponent.cs:10]: NetworkedComponent prolly.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

**Review by metalgearsloth — APPROVED** [review]
